cpp/python/src/python_future_task_collector.cpp

```diff
@@ -0,0 +1,54 @@
+/**
+ * SPDX-FileCopyrightText: Copyright (c) 2023, NVIDIA CORPORATION & AFFILIATES.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include <mutex>
+
+#include <Python.h>
+
+#include <ucxx/log.h>
+#include <ucxx/python/python_future_task_collector.h>
+
+namespace ucxx {
+
+namespace python {
+
+PythonFutureTaskCollector& PythonFutureTaskCollector::get()
+{
+  static PythonFutureTaskCollector collector;
+  return collector;
+}
+
+void PythonFutureTaskCollector::push(PyObject* handle)
+{
+  std::lock_guard<std::mutex> lock(_mutex);
+  _toCollect.push_back(handle);
+}
+
+void PythonFutureTaskCollector::collect()
+{
+  PyGILState_STATE state = PyGILState_Ensure();
+
+  {
+    std::lock_guard<std::mutex> lock(_mutex);
+    for (auto& handle : _toCollect)
+      Py_XDECREF(handle);
+    ucxx_trace("Collected %lu PythonFutureTasks", _toCollect.size());
+    _toCollect.clear();
+  }
+
+  PyGILState_Release(state);
+}
+
+PythonFutureTaskCollector::PythonFutureTaskCollector() {}
+
+PythonFutureTaskCollector::~PythonFutureTaskCollector()
+{
+  if (_toCollect.size() > 0)
+    ucxx_warn("Destroying PythonFutureTaskCollector with %lu uncollected tasks", _toCollect.size());
```

Comment:
**suggestion:** this should hold the lock to be race-free I think.

Reply:
Done via https://github.com/rapidsai/ucxx/pull/46/commits/7e0ad1b7de36915eb3edf78898e1208e5eebcca6 .